File: Plugin/src/main/java/xyz/hstudio/horizon/module/VerticalMovement.java
package xyz.hstudio.horizon.module;

import xyz.hstudio.horizon.HPlayer;
import xyz.hstudio.horizon.event.InEvent;
import xyz.hstudio.horizon.event.inbound.MoveEvent;
import xyz.hstudio.horizon.util.enums.Direction;

/**
 * Vertical movement check.
 * <p>
 * a = (vt - v0) / t -> vt = v0 + at
 * <p>
 * vt is the velocity of the time t
 * v0 is the velocity of the beginning
 * a is the gravitational acceleration (g = -0.08 in Minecraft)
 * t is the time, measured in ticks
 * <p>
 * we also need to multiply it with vertical air resistance (k = 0.98 in Minecraft)
 * <p>
 * so expectedYVelocity = (prevYVelocity + -0.08 * 1) * 0.98
 */
public class VerticalMovement extends CheckBase {

    private float estimatedYVelocity;

    public VerticalMovement(HPlayer p) {
        super(p, 1, 10, 10);
    }

    @Override
    public void received(InEvent<?> event) {
        if (event instanceof MoveEvent) {
            move((MoveEvent) event);
        }
    }

    private void move(MoveEvent e) {
        float deltaY = (float) e.velocity.y;
        float estimatedYVelocity = (this.estimatedYVelocity + -0.08F) * 0.98F;
        boolean skip = false;
        if (e.step || e.teleport) {
            estimatedYVelocity = 0.0F;
            skip = true;
        } else if (e.jump || e.knockBack || e.onSlime) {
            estimatedYVelocity = deltaY;
            skip = true;
        } else if (e.onGround && !this.p.physics.onGround) {
            estimatedYVelocity = 0F;
            skip = true;
        } else if (Math.abs(estimatedYVelocity) < 0.005 && deltaY <= 0 && p.physics.velocity.y > 0) {
            estimatedYVelocity = deltaY == 0 ? 0 : -(Math.min(Math.abs(deltaY - -0.0784F), Math.abs(deltaY - -0.0754F)) - deltaY);
            skip = true;
        }
        // Blocks above
        boolean hitHead = e.touchedFaces.contains(Direction.UP);
        boolean hasHitHead = p.physics.touchedFaces.contains(Direction.UP);
        if (hitHead && !hasHitHead) {
            deltaY = estimatedYVelocity = 0;
        }

        if (!skip) {
            if (e.onGround) {
                estimatedYVelocity = 0F;
            }

            float error = Math.abs(deltaY - estimatedYVelocity);
            if (error > 0.001) {
                System.out.println("Invalid motion! Delta:" + deltaY + ", Estimated:" + estimatedYVelocity + ", Error:" + error);
            }
        }
        this.estimatedYVelocity = estimatedYVelocity;
    }
}

File: Plugin/src/main/java/xyz/hstudio/horizon/module/VerticalMovement.kt
package xyz.hstudio.horizon.module

import xyz.hstudio.horizon.HPlayer
import xyz.hstudio.horizon.event.InEvent
import xyz.hstudio.horizon.event.inbound.MoveEvent
import xyz.hstudio.horizon.util.Physics.AIR_RESISTANCE_VERTICAL
import xyz.hstudio.horizon.util.Physics.GRAVITATIONAL_ACCELERATION
import kotlin.math.abs

/**
 * Vertical movement check.
 * <p>
 * a = (vt - v0) / t -> vt = v0 + at
 * <p>
 * vt is the velocity of the time t
 * v0 is the velocity of the beginning
 * a is the gravitational acceleration (g = -0.08 in Minecraft)
 * t is the time, measured in ticks
 * <p>
 * we also need to multiply it with vertical air resistance (k = 0.98 in Minecraft)
 *
 * @see xyz.hstudio.horizon.util.Physics
 * so expectedYVelocity = (prevYVelocity + -0.08 * 1) * 0.98
 */
class VerticalMovement(p: HPlayer) : CheckBase(p, 1, 10, 10) {

    private var estimatedYVelocity = 0f

    override fun received(event: InEvent?) {
        if (event is MoveEvent) {
            move(event)
        }
    }

    private fun move(e: MoveEvent) {
        val deltaY = e.velocity.y.toFloat()
        var estimatedYVelocity = (estimatedYVelocity + GRAVITATIONAL_ACCELERATION) * AIR_RESISTANCE_VERTICAL
        var skip = false
        when {
            e.step -> {
                estimatedYVelocity = 0f
                skip = true
            }
            e.jump -> {
                estimatedYVelocity = deltaY
                skip = true
            }
            e.onGround && !p.physics.onGround -> {
                estimatedYVelocity = deltaY
                skip = true
            }
        }
        if (!skip) {
            if (e.onGround) {
                estimatedYVelocity = 0f
            } else {
                // Fix jump bug when standing still
                // Wtf MoJang???
                if (abs(estimatedYVelocity) < 0.005 && deltaY < 0 && p.physics.velocity.y > 0 && e.to.distance2dSquared(p.physics.position) < 1E-6) {
                    estimatedYVelocity = deltaY
                }
                /*
                // Fix 1.8 jump bug
                if (abs(estimatedYVelocity) < 0.003 && estimatedYVelocity != 0f && deltaY <= 0) {
                    estimatedYVelocity = deltaY
                }
                // Fix 1.9 jump bug
                if (deltaY < 0 && p.physics.prevVelocity.y >= 0 && e.to.distance2dSquared(p.physics.position) < 1E-6) {
                    estimatedYVelocity = deltaY
                }
                */
            }

            // Keep four decimal places
            val error = ((deltaY - estimatedYVelocity) * 10000).toInt() / 10000f
            if (error > 0.001) {
                println("Error:$error")
            }
        }

        this.estimatedYVelocity = estimatedYVelocity
    }
}